### PR TITLE
feat: adds support for esbuild plugins (SC-2514)

### PR DESCRIPTION
**Pull request by eahefnawy**
Closes https://github.com/serverless/serverless/issues/12654

### CODE REVIEW

File: docs/providers/aws/guide/building.md [hunk 1]
```diff
@@ -46,6 +46,31 @@ build:
...
+
+The JavaScript file must export a function that returns an esbuild configuration object. For your convenience, the **serverless** instance is passed to that function.
+
+Here's an example using the `esbuild-plugin-env` plugin:
+
+```js
+// don't forget to install this package
+const env = require('esbuild-plugin-env')
+
+module.exports = (serverless) => {
```
Comment (Mmarzex) [inline docs/providers/aws/guide/building.md:66]: We should include an ESM example as well.

File: lib/plugins/esbuild/index.js [hunk 2]
```diff
@@ -9,6 +10,7 @@ import _ from 'lodash'
 import pLimit from 'p-limit'
 import globby from 'globby'
 import ServerlessError from '../../serverless-error.js'
+import { config } from 'dotenv'
```
Comment (Mmarzex) [inline lib/plugins/esbuild/index.js:13]: I don't think this is used.

File: lib/plugins/esbuild/index.js [hunk 3]
```diff
@@ -273,15 +275,47 @@ class Esbuild {
     return undefined
   }
 
-  _buildProperties() {
+  async _buildProperties() {
```
Comment (Mmarzex) [inline lib/plugins/esbuild/index.js:278]: Can we add some sort of memoization/caching of this function? Right now we'd be importing that module multiple times per run when we should only be doing the import once.

**Review by Mmarzex — CHANGES_REQUESTED** [review]

File: lib/plugins/esbuild/index.js (hunk 3 above)
Comment (eahefnawy) [reply lib/plugins/esbuild/index.js:278]: Hmmm wouldn't that have some side effects in dev mode though? If I make changes to that file while in an active dev mode session, these changes won't be picked up due to the memoization.

Comment (Mmarzex) [reply lib/plugins/esbuild/index.js:278]: We could have different behavior for dev mode. I'm just concerned that we are re-importing potentially large files for no reason in these cases.

Comment (eahefnawy) [reply lib/plugins/esbuild/index.js:278]: I understand. Ok I'll memoize it.

**Review by Mmarzex — APPROVED** [review]

**Pull request merged**